data/appRateLimiter.go

```diff
@@ -0,0 +1,267 @@
+		maxBucketSize = 1
+	}
+	r := &appRateLimiter{
+		maxBucketSize:        maxBucketSize,
+		serviceRatePerWindow: serviceRatePerWindow,
+		serviceRateWindow:    serviceRateWindow,
+		seed:                 crypto.RandUint64(),
+	}
+	crypto.RandBytes(r.salt[:])
+
+	for i := 0; i < numBuckets; i++ {
+		r.buckets[i] = make(map[keyType]*appRateLimiterEntry)
+		r.lrus[i] = util.NewList[keyType]().AllocateFreeNodes(maxBucketSize)
+	}
+	return r
+}
+
+func (r *appRateLimiter) entry(b int, key keyType, curInt int64) (*appRateLimiterEntry, bool) {
+	r.mus[b].Lock()
+	defer r.mus[b].Unlock()
+
+	if len(r.buckets[b]) >= r.maxBucketSize {
+		// evict the oldest entry
+		start := time.Now()
+
+		el := r.lrus[b].Back()
+		delete(r.buckets[b], el.Value)
+		r.lrus[b].Remove(el)
+
+		atomic.AddUint64(&r.evictions, 1)
+		atomic.AddUint64(&r.evictionTime, uint64(time.Since(start)))
+	}
+
+	entry, ok := r.buckets[b][key]
+	if ok {
+		el := entry.lruElement
+		// note, the entry is marked as recently used even before the rate limiting decision
+		// since it does not make sense to evict keys that are actively attempted
+		r.lrus[b].MoveToFront(el)
+
+		// the same logic is applicable to the intervals: if a new interval is started, update the entry
+		// by moving the current value to the previous and resetting the current.
+		// this is done under a lock so that the interval is not updated concurrently.
+		// The rationale is even this requests is going to be dropped the new interval already started
+		// and it is OK to start a new interval and have it prepared for upcoming requests
+		var newPrev int64 = 0
+		switch entry.interval {
+		case curInt:
+			// the interval is the same, do nothing
+		case curInt - 1:
+			// these are continuous intervals, use current value as a new previous
+			newPrev = entry.cur.Load()
+			fallthrough
+		default:
+			// non-contiguous intervals, reset the entry
+			entry.prev.Store(newPrev)
+			entry.cur.Store(0)
+			entry.interval = curInt
+		}
+	} else {
+		el := r.lrus[b].PushFront(key)
+		entry = &appRateLimiterEntry{interval: curInt, lruElement: el}
+		r.buckets[b][key] = entry
+	}
+	return entry, ok
+}
+
+// interval calculates the interval numeric representation based on the given time
+func (r *appRateLimiter) interval(now time.Time) int64 {
+	return now.UnixNano() / int64(r.serviceRateWindow)
+}
+
+// fraction calculates the fraction of the interval that is elapsed since the given time
+func (r *appRateLimiter) fraction(now time.Time) float64 {
+	return float64(now.UnixNano()%int64(r.serviceRateWindow)) / float64(r.serviceRateWindow)
+}
+
+// shouldDrop returns true if the given transaction group should be dropped based on the
+// on the rate for the applications in the group: the entire group is dropped if a single application
+// exceeds the rate.
+func (r *appRateLimiter) shouldDrop(txgroup []transactions.SignedTxn, origin []byte) bool {
+	return r.shouldDropInner(txgroup, origin, time.Now())
```

Comment:
Can we avoid the word "Inner", since we're talking about transaction and it's easy to wonder what "inner transactions" have to do with this.

How about `shouldDropAt()`?